apiserver/facades/client/application/application.go

```diff
@@ -451,26 +452,62 @@ func (api *APIBase) Deploy(args params.ApplicationsDeploy) (params.ErrorResults,
 }
 
 // compatibilityApplicationDeployArgs ensures that Deploy calls from
-// a juju 3.x client will work against a juju 2.9.x controller. In
-// juju 3.x, params.ApplicationsDeploy was changed to remove series
+// clients such as juju 3.x, python-libjuju 3.x, etc. work against a 2.9.x
+// controller. To support these clients, we make no assumptions about
+// which of the 3 potential places the OS is specified. We check all 3,
+// assert any provided match, and fill in the remain series.
+//
+// In juju 3.x, params.ApplicationsDeploy was changed to remove series
 // however, the facade version was not changed, nor was the name of
 // the params.ApplicationsDeploy changed. Thus it appears you can use
 // a juju 3.x client to deploy from a juju 2.9 controller, however it
-// fails because the series was not found. Make those corrections here.
+// fails because the series was not found.
+//
+// NOTE: We do not mess with charm url because this is dangerous. Mutating
+// charm url can lead to unexpected results
 func compatibilityApplicationDeployArgs(arg params.ApplicationDeploy) (params.ApplicationDeploy, error) {
-	origin := arg.CharmOrigin
-	if origin.Series != "" {
-		return arg, nil
+	s, err := getUnifiedSeries(arg)
+	if err != nil {
+		return arg, errors.Annotatef(err, "cannot deploy application")
 	}
-	originSeries, err := series.GetSeriesFromChannel(origin.Base.Name, origin.Base.Channel)
+	b, err := series.GetBaseFromSeries(s)
 	if err != nil {
-		return arg, err
+		return arg, errors.Trace(err)
 	}
-	origin.Series = originSeries
-	arg.Series = originSeries
+
+	arg.Series = s
+	arg.CharmOrigin.Series = s
+	arg.CharmOrigin.Base.Name = b.Name
+	arg.CharmOrigin.Base.Channel = b.Channel.String()
+
 	return arg, nil
 }
 
+// getUnifiedSeries checks the multiple places a series can be provided in a params.ApplicationDeploy,
+// checks they match if they're provided, and returns the matching series
+func getUnifiedSeries(arg params.ApplicationDeploy) (string, error) {
+	argSeries := arg.Series
+	originSeries := arg.CharmOrigin.Series
+	originBaseSeries, _ := series.GetSeriesFromChannel(arg.CharmOrigin.Base.Name, arg.CharmOrigin.Base.Channel)
+
+	s := argSeries
+	if s == "" {
+		s = originSeries
+	}
+	if s == "" {
+		s = originBaseSeries
+	}
+	if s == "" {
+		return "", errors.Errorf("no series provided")
+	}
+	if (argSeries != "" && s != argSeries) || (originSeries != "" && s != originSeries) || (originBaseSeries != "" && s != originBaseSeries) {
```

Comment:
This is a confusing line with a heavy logic. If I understand this right, this is just a validation check, we have no problem if one or more of `argSeries`, `originSeries` and `originBaseSeries` are provided, as long as whatever provided all have the same value, is that right?

If that's the case, then I'd suggest 

1) definitely write a clear comment on what's going on, 
2) change the error message down in line 506 to something like:

```
return "", errors.Errorf("inconsistent values for series detected. argument: %q, charm origin: %q, charm base: %q", argSeries, originSeries, originBaseSeries)
```

Reply:
This logic has been re-written now that I'm including the charm url series, which hopefully clarifies things.

I have made these changes as well anyway